std/algorithm/iteration.d

```diff
@@ -1502,9 +1502,12 @@ if (isInputRange!R)
     static if (isForwardRange!R)
     {
         ///
-        @property typeof(this) save() {
+        @property typeof(this) save()
+        {
+            import std.algorithm.mutation : move;
             typeof(this) ret = this;
-            ret._input = this._input.save;
+            auto saved = this._input.save;
+            move(saved, ret._input);
```

Comment:
>I'm not sure if this is the right direction.

I'm not sure either. If we're doing it this way though we could at least put this in a branch that checks for `hasElaborateAssign` and otherwise use the normal behavior.

Reply:
> we could at least put this in a branch that checks for hasElaborateAssign and otherwise use the normal behavior.

What would that accomplish? It looks like `move` boils down to `target = source;` anyway when there's nothing special to be done.

https://github.com/dlang/phobos/blob/5b5e5281a88d24fd31fa316c17850c960ae41664/std/algorithm/mutation.d#L1376-L1379

Reply:
>What would that accomplish?

It avoids unnecessary function calls and unnecessary data copying and potentially reduces template bloat. Compare the generated assembly for `aaaCopyImplAAA` and `bbbCopyImplBBB` with compile options "-O -inline": https://run.dlang.io/is/1aofWf
```d
struct Inner
{
    uint meaninglessField;
    @property typeof(this) save() { return this; }
}

struct ExampleStruct
{
    Inner exampleField;
    @property typeof(this) aaaCopyImplAAA()
    {
        typeof(this) ret = this;
        ret.exampleField = this.exampleField.save;
        return ret;
    }
    @property typeof(this) bbbCopyImplBBB()
    {
        import std.algorithm.mutation : move;
        typeof(this) ret = this;
        auto saved = this.exampleField.save;
        move(saved, ret.exampleField);
        return ret;
    }        
}

void main()
{
}
```
aaaCopyImplAAA:
```
onlineapp.ExampleStruct.aaaCopyImplAAA():
		push	RBP
		mov	RBP,RSP
		mov	EAX,[RDI]
		pop	RBP
		ret
```
bbbCopyImplBBB:
```
onlineapp.ExampleStruct.bbbCopyImplBBB():
		push	RBP
		mov	RBP,RSP
		sub	RSP,010h
		mov	EAX,[RDI]
		mov	-8[RBP],EAX
		mov	ECX,[RDI]
		mov	-4[RBP],ECX
		lea	RSI,-4[RBP]
		lea	RDI,-8[RBP]
		call	  pure nothrow @nogc void std.algorithm.mutation.moveImpl!(onlineapp.Inner).moveImpl(ref onlineapp.Inner, ref onlineapp.Inner)@PLT32
		mov	EAX,-8[RBP]
		mov	RSP,RBP
		pop	RBP
		ret
		add	[RCX],AH

pure nothrow @nogc void std.algorithm.mutation.moveImpl!(onlineapp.Inner).moveImpl(ref onlineapp.Inner, ref onlineapp.Inner):
		push	RBP
		mov	RBP,RSP
		sub	RSP,010h
		mov	-010h[RBP],RBX
		mov	RCX,RDI
		mov	RDX,RSI
		cmp	RDX,RCX
		jne	L20
		mov	RBX,-010h[RBP]
		mov	RSP,RBP
		pop	RBP
		ret
L20:		cmp	RDX,RCX
		je	L32
		mov	EBX,[RDX]
		mov	[RCX],EBX
		mov	RBX,-010h[RBP]
		mov	RSP,RBP
		pop	RBP
		ret
L32:		mov	R8D,055Fh
		lea	RCX,_TMP0@PC32[RIP]
		mov	EAX,040h
		mov	RDX,RAX
		mov	-8[RBP],RDX
		lea	RDX,_TMP0@PC32[RIP]
		mov	EDI,027h
		mov	RSI,RDX
		mov	RDX,-8[RBP]
		call	  _d_assert_msg@PLT32
		mov	RBX,-010h[RBP]
		mov	RSP,RBP
		pop	RBP
		ret
```

Reply:
Always compare ASM from LDC and never from DMD unless you're working on DMD's backend. DMD's optimizer is out of date, and therefore should not be catered to.

Reply:
Yes, please don't "optimize" code based on what dmd does.  Due to dmd's inferior optimizer, this tends to greatly uglify code needlessly, where ldc/gdc would produce code better than dmd anyway, even when written in the original "unoptimized" form.

We should rather improve dmd's optimizer than make Phobos a rats' nest of workarounds for it.

Reply:
Gotcha, and LDC handles this example nicely. (The output is still littered with unused functions produced by template expansion, but I believe that's being actively worked on.)

Reply:
> The output is still littered with unused functions produced by template expansion

`-L--gc-sections` should be able to remove unused sections.
(see e.g. https://sourceware.org/binutils/docs/ld/Options.html)